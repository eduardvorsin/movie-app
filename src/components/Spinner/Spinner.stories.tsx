import type { Meta, StoryObj } from '@storybook/react';
import Spinner, { Props } from './Spinner';

type Sizes = Exclude<Props['size'], undefined>;
const sizes: Sizes[] = ['small', 'medium', 'large', 'xlarge'];

type Variant = {
	size: Sizes,
	color: Props['color'],
}

const variants: Variant[] = [
	{
		size: 'small',
		color: 'stroke-lime-800 dark:stroke-lime-300',
	},
	{
		size: 'medium',
		color: 'stroke-red-800 dark:stroke-red-300',
	},
	{
		size: 'large',
		color: 'stroke-orange-800 dark:stroke-orange-300',
	},
	{
		size: 'xlarge',
		color: 'stroke-teal-800 dark:stroke-teal-300',
	},
];

const meta: Meta<typeof Spinner> = {
	title: 'components/Spinner',
	tags: ['autodocs'],
	argTypes: {
		size: {
			description: 'Sets the size of the spinner. You can choose a ready-made size or set your own styles for sizes.',
		},
		className: {
			description: 'Additional classes for the Spinner'
		},
		color: {
			description: 'Color for spinner',
		},
		testId: {
			description: 'Id for testing the component',
		}
	},
	component: Spinner,
	parameters: {
		docs: {
			description: {
				component: 'Spinners serve the purpose of indicating a loading condition to the user and can be employed in buttons, forms, containers, pages, or any other context requiring the representation of a loading state.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof Spinner>;

export const Default: Story = {
	parameters: {
		docs: {
			description: {
				story: 'Default type of spinner',
			},
		},
	},
};

export const Color: Story = {
	args: {
		color: 'stroke-lime-500 dark:stroke-lime-300',
	},
	parameters: {
		docs: {
			description: {
				story: 'The color can be changed if you pass to the prop color class with a new color for svg stroke',
			},
		},
	},
};

export const Sizes: Story = {
	parameters: {
		docs: {
			description: {
				story: 'Allows you to set the size of the spinner. The default is medium. There are 5 spinner sizes in total, of which 4 have specific width and height dimensions and the 5th is custom, allowing you to set your sizes through classes',
			},
		},
	},
	render: () => (
		<div
			className='flex items-start'
		>
			{sizes.map((size) => (
				<Spinner
					className='mr-4 last:mr-0'
					key={size}
					size={size}
				/>
			))}
		</div>
	)
};

export const All: Story = {
	parameters: {
		docs: {
			description: {
				story: 'Here are all kinds of spinners, in size and colors',
			},
		},
	},
	render: () => (
		<div
			className='flex items-start'
		>
			{variants.map((variant) => (
				<Spinner
					className='mr-4 last:mr-0'
					key={variant.size}
					size={variant.size}
					color={variant.color}
				/>
			))}
		</div>
	)
};