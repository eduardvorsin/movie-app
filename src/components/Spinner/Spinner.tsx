type Props = {
	className?: string,
	size?: 'small' | 'medium' | 'large' | 'xlarge' | 'custom',
	testId?: string,
	color?: string,
};

const sizesTypes = {
	small: 'w-4, h-4',
	medium: 'w-6 h-6',
	large: 'w-12 h-12',
	xlarge: 'w-24 h-24',
	custom: '',
} as const;

export default function Spinner({
	className,
	size = 'medium',
	color,
	testId,
}: Props) {
	const classes = [
		'inline-flex align-middle',
		sizesTypes[size],
		className,
	].join(' ');

	const circleClasses = [
		'stroke-[0.75] fill-none',
		color ? color : 'dark:stroke-blue-300 stroke-blue-800',
	].join(' ');

	return (
		<span
			className={classes}
			data-testid={testId}
		>
			<svg
				className='animate-spin'
				viewBox='0 0 16 16'
			>
				<circle
					className={circleClasses}
					cx='8'
					cy='8'
					r='7'
					strokeDasharray={60}
					strokeDashoffset={50}
					strokeLinecap='round'
				>
				</circle>
			</svg>
		</span>
	);
};