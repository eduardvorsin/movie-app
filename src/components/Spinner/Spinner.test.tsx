import { render, screen } from '@testing-library/react';
import Spinner from './Spinner';

describe('Spinner tests', () => {
	it('is rendered correctly', () => {
		render(<Spinner testId='spinner' />);

		expect(screen.getByTestId<HTMLSpanElement>('spinner')).toBeInTheDocument();
	});

	it('if the size is xlarge, then classes w-24 and h-24 should be added', () => {
		render(<Spinner testId='spinner' size='xlarge' />);

		expect(screen.getByTestId<HTMLSpanElement>('spinner')).toHaveClass('w-24 h-24')
	});

	it('is a basic snapshot', () => {
		render(<Spinner testId='spinner' />);

		expect(screen.getByTestId<HTMLSpanElement>('spinner')).toMatchSnapshot();
	});

	it('is a snapshot with a size of type "small"', () => {
		render(<Spinner testId='spinner' size='small' />);

		expect(screen.getByTestId<HTMLSpanElement>('spinner')).toMatchSnapshot();
	});

	it('is a snapshot with a size of type "medium"', () => {
		render(<Spinner testId='spinner' size='medium' />);

		expect(screen.getByTestId<HTMLSpanElement>('spinner')).toMatchSnapshot();
	});

	it('is a snapshot with a size of type "large"', () => {
		render(<Spinner testId='spinner' size='large' />);

		expect(screen.getByTestId<HTMLSpanElement>('spinner')).toMatchSnapshot();
	});

	it('is a snapshot with a size of type "xlarge"', () => {
		render(<Spinner testId='spinner' size='xlarge' />);

		expect(screen.getByTestId<HTMLSpanElement>('spinner')).toMatchSnapshot();
	});
});