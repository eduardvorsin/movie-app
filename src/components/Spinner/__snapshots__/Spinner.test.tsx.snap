// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner tests is a basic snapshot 1`] = `
<span
  class="inline-flex align-middle w-6 h-6 "
  data-testid="spinner"
>
  <svg
    class="animate-spin"
    viewBox="0 0 16 16"
  >
    <circle
      class="stroke-[0.75] fill-none dark:stroke-blue-300 stroke-blue-800"
      cx="8"
      cy="8"
      r="7"
      stroke-dasharray="60"
      stroke-dashoffset="50"
      stroke-linecap="round"
    />
  </svg>
</span>
`;

exports[`Spinner tests is a snapshot with a size of type "custom" 1`] = `
<span
  class="inline-flex align-middle  "
  data-testid="spinner"
>
  <svg
    class="animate-spin"
    viewBox="0 0 16 16"
  >
    <circle
      class="stroke-[0.75] fill-none dark:stroke-blue-300 stroke-blue-800"
      cx="8"
      cy="8"
      r="7"
      stroke-dasharray="60"
      stroke-dashoffset="50"
      stroke-linecap="round"
    />
  </svg>
</span>
`;

exports[`Spinner tests is a snapshot with a size of type "large" 1`] = `
<span
  class="inline-flex align-middle w-12 h-12 "
  data-testid="spinner"
>
  <svg
    class="animate-spin"
    viewBox="0 0 16 16"
  >
    <circle
      class="stroke-[0.75] fill-none dark:stroke-blue-300 stroke-blue-800"
      cx="8"
      cy="8"
      r="7"
      stroke-dasharray="60"
      stroke-dashoffset="50"
      stroke-linecap="round"
    />
  </svg>
</span>
`;

exports[`Spinner tests is a snapshot with a size of type "medium" 1`] = `
<span
  class="inline-flex align-middle w-6 h-6 "
  data-testid="spinner"
>
  <svg
    class="animate-spin"
    viewBox="0 0 16 16"
  >
    <circle
      class="stroke-[0.75] fill-none dark:stroke-blue-300 stroke-blue-800"
      cx="8"
      cy="8"
      r="7"
      stroke-dasharray="60"
      stroke-dashoffset="50"
      stroke-linecap="round"
    />
  </svg>
</span>
`;

exports[`Spinner tests is a snapshot with a size of type "small" 1`] = `
<span
  class="inline-flex align-middle w-4, h-4 "
  data-testid="spinner"
>
  <svg
    class="animate-spin"
    viewBox="0 0 16 16"
  >
    <circle
      class="stroke-[0.75] fill-none dark:stroke-blue-300 stroke-blue-800"
      cx="8"
      cy="8"
      r="7"
      stroke-dasharray="60"
      stroke-dashoffset="50"
      stroke-linecap="round"
    />
  </svg>
</span>
`;

exports[`Spinner tests is a snapshot with a size of type "xlarge" 1`] = `
<span
  class="inline-flex align-middle w-24 h-24 "
  data-testid="spinner"
>
  <svg
    class="animate-spin"
    viewBox="0 0 16 16"
  >
    <circle
      class="stroke-[0.75] fill-none dark:stroke-blue-300 stroke-blue-800"
      cx="8"
      cy="8"
      r="7"
      stroke-dasharray="60"
      stroke-dashoffset="50"
      stroke-linecap="round"
    />
  </svg>
</span>
`;
